### DISCUSSION TITLE
How to get reference to native android activity?

### DISCUSSION THREAD

**Discussion by MaxwellDAssistek** [AvaloniaUI/Avalonia#17350] [Q&A]
Hello,

I need to get a reference to the native android that Avalonia is running in to do some native android manipulation. Is there any way to get at it through Avalonia without resorting to just storing a reference in a random static variable.

Poking around in the debugger shows that `Application.Current.ApplicationLifetime.Activity` exists but it is in the `Avalonia.Android.SingleViewLifetime` class which is marked internal. I guess I could use reflection to get at it, but that feels dirty.

Thank you!

**Comment by maxkatz6** [comment]
> Is there any way to get at it through Avalonia without resorting to just storing a reference in a random static variable.

From what I know, that's a typical recommendation for native android development. There are no android APIs to get current activity. Avalonia doesn't add one either.

Keep in mind, that activity can be re-created in the same process multiple times. If android OS suspends app and restores it. Normally, it would execute OnCreate method each time.

**Reply by MaxwellDAssistek** [reply]
In my specific case, the problem is that I need to access it from a module that cannot reference the base project, so my two choices are either very nasty reflection to get the base assembly and locate that static variable or reflection to get at `Application.Current.ApplicationLifetime.Activity`. I'm going to go with the latter. 

 I understand that mobile platforms are somewhat new to Avalonia, but there is much to improve still in the native flexibility department. In MAUI this is accessible through `Microsoft.Maui.ApplicationModel.Platform.CurrentActivity`.

**Reply by maxkatz6** [reply]
We try to respect platform specific limitations, when they occur. And not having CurrentActivity is one of these.

I.e. the big problem of API design by adding CurrentActivity static in the framework is - users won't understand, when this property is invalidated. They will read it once, cache it and expect this activity to be always correct. Instead, MAUI has to assume that developer will always re-read this property, returning up-to-date activity instance each time.

**Reply by maxkatz6** [reply]
> In my specific case, the problem is that I need to access it from a module that cannot reference the base project

You can work around this relatively easily. Better than using reflection.
I can see two options right away, but there might be more:
1. Add `object NativeApplication {get; set;}` property in the Avalonia Application, which is set from OnCreate, but accessible from shared projects.
2. Add "net8.0-android" target framework to the shared project, and separated code by "#if ANDROID" (just like in MAUI, it's a common feature in .NET SDK). This way you can access APIs from Android in the shared project.

**Reply by MaxwellDAssistek** [reply]
That sounds like a pretty arbitrary line to draw. I believe that MAUI is making a pretty safe assumption that could simply be mentioned in the documentation. In any case, putting this on the developer to keep track of seems much more error prone due to developers not understanding all the nuances of when the static variable they make is to be updated or invalidated.

**Reply by MaxwellDAssistek** [reply]
The workaround you mentioned doesn't work for me. The code that needs the reference to the activity is in a class library project (that already has platform specific code) that is loaded dynamically without referencing any of the base application projects.
